### DISCUSSION TITLE
executing an async js method on the client

### DISCUSSION THREAD

**Discussion by shaharzi** [cefsharp/CefSharp#3733] [General]
Hello,

Im new in cefsharp, im trying to execute an async method on our client with no success on the result. 

example of code trying to execute:

` window.query = async (args) => {
                const response = await fetch('https://jsonplaceholder.typicode.com/todos/1');
                const todo =  await response.json();
                return todo;
        };`

anything im trying from c#, the result not return.

any help with that?

thanks !

**Comment by amaitland** [comment]
You need to use EvaluateScriptAsPromiseAsync see https://github.com/cefsharp/CefSharp/wiki/General-Usage#2-how-do-you-call-a-javascript-method-that-returns-a-result

**Reply by shaharzi** [reply]
thanks, but i don't understand the example over there. the EvaluateScriptAsPromiseAsync wants me to inject script from c# but our client has is own async function. so something im missing here.

**Reply by amaitland** [reply]
If you simply need to execute an existing JavaScript function then execute JavaScript that calls the method.

**Reply by shaharzi** [reply]
i did that, but the client defined his function like above, with async, and the response does not arrived, arrive null. i need to get response. 

i executed this: 

` public async Task<JavascriptResponse> EvaluateMacefScriptAsync(dynamic request, string methodName){
            return await browser.EvaluateScriptAsync(methodName, new object[] { JsonConvert.SerializeObject(request) });
        }`

**Reply by amaitland** [reply]
As you need to get the result of a promise you need to use EvaluateScriptAsPromiseAsync

**Reply by shaharzi** [reply]
i'll be more specific. i want from c# to call a js method and get result. but the js method is with async.

**Reply by ukandrewc** [reply]
I'd be interested to see what @amaitland says, but I'd suggest that you can't do that.

You'd probably need to use an async JS wrapper with an await followed by a call back to your C#

**Reply by shaharzi** [reply]
tried that and its not working

**Reply by ukandrewc** [reply]
I'm using that to return result of async functions, like this:

in .Net
```
Browser.ExecuteScriptAsync('wrapper()')

function myCallBack(result){
}
```

in JS:
```
async function wrapper(){
  r = await asyncFunction()
  MyObject.myCallBack(r)
}
```

**Reply by amaitland** [reply]
> i'll be more specific. i want from c# to call a js method and get result. but the js method is with async.

Fundimendally an async function returns a promise. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#return_value

> but I'd suggest that you can't do that.

@ukandrewc You can using EvaluateScriptAsPromiseAsync

https://github.com/cefsharp/CefSharp/blob/4ef267cabf65cc84753afc1f7bd7be0bfdf65dd1/CefSharp.Test/OffScreen/OffScreenBrowserBasicFacts.cs#L294

> tried that and its not working

@shaharzi If you've made an attempt then you need to provide an example of what you have tried.

**Reply by ukandrewc** [reply]
@amaitland Thanks, I'll have a look at that, as it may be useful - still lots to learn ;-)

**Reply by shaharzi** [reply]
Eventually i used this approach and it worked 

` public async Task<JavascriptResponse> EvaluateMacefScriptAsync(dynamic request, string methodName)
        {
            var script = WebBrowserExtensions.GetScriptForJavascriptMethodWithArgs(methodName,
                new object[] {JsonConvert.SerializeObject(request)});

            var temp = await browser.EvaluateScriptAsPromiseAsync($"return await {script};");

            return temp;
        }`
